datafusion/functions/src/encoding/inner.rs

```diff
@@ -49,17 +48,8 @@ impl Default for EncodeFunc {
 
 impl EncodeFunc {
     pub fn new() -> Self {
-        use DataType::*;
         Self {
-            signature: Signature::one_of(
```

Comment:
It seems to me that moving the signature from a data driven description (aka describe "what" is needed and letting some other code compute if the given arguments match that signature), this PR is moving many of the functions towards more functional (each function has to implement its own custom coercion, likely resulting in significant duplication).

What do you think (perhaps as a follow on PR) of adding `DataType::Null` support to the Signature calculations somehow rather than inlining / duplicating the coercion logic?

Maybe something like 
```
Signature::allow_null(..)
```
that would support automatically coercing arguments from null?

Or maybe we should always support coercing Null to any type

Reply:
Alternative signature like Signature::String, similar to Signature::numeric that includes converting null to string too?

Reply:
I am not sure -- I was just reacting that this "handle null" pattern seems common and it seems like this approach will require custom coerce logic for all functions 🤔

Reply:
Null to T coercion needs to be handled elsewhere anyway (eg when computing type of a UNION, etc.).
We can free functions from having to bother about coercions at all and let the engine calculate coercions when building the logical plan.

This is actually super fundamental for DataFusion vision as a composable query engine. Coercion rules are very implementation-specific. If we had functions spiced up with coercions inside them, that would make those functions non-reusable.

cc @wizardxz @sadboy

Reply:
100%

It seems to me like `Signature` is supposed to communicate what types the function implementation has a native implementation for and the coercion of whatever the user provided doesn't match one of the supported types

Reply:
@findepi Are you suggesting something like general coercion that is non-function specific? But what if we want different coercion rule for different function, we might need to do coercion function wise

Reply:
Why would we want different coercion rules for different functions?

Reply:
My idea is that it is more flexible to the user, although, without the real use case, it might be a premature optimization 🤔.

Reply:
Sorry to chip in late; this PR addresses other issues, such as #12307. I wonder if I could split it and leave the changes regarding the coercion of functions in this one (to keep the discussion in one place) and the others in a new PR. Would that be ok?